### DISCUSSION TITLE
Lootbox v1.0.0 - Cashback Rewards

### DISCUSSION THREAD

**Discussion by 0x4007** [pavlovcik/uad-ui#158] [Protocol Specifications]
# v1.0.0 - Proof of concept - Cashback on ANY NFT Purchases on OpenSea

We want to increase usage of the uAD by giving people a reason to purchase NFTs with uAD. Introducting Ubiquity Credit Cashback Program, similar to credit cards!

+ #158 
+ #165 
+ #166 

## Overview

> アレクサンダー.eth | U B I Q U I T Y:
i came up with a hack to make the opensea thing work

> アレクサンダー.eth | U B I Q U I T Y:
> <img width="1552" alt="Screen Shot 2022-05-13 at 05 26 17" src="https://user-images.githubusercontent.com/4975670/168184188-5c33ed1f-1a11-4595-a17c-3103d5cf6bb0.png">

> アレクサンダー.eth | U B I Q U I T Y:
so basically we modify our linkedin chrome extension (already has great infra, client side!) to modify the opensea ui

> アレクサンダー.eth | U B I Q U I T Y:
then we deploy one or two contracts to middle man the transaction

> アレクサンダー.eth | U B I Q U I T Y:
in these contracts, we take, for example, a 2.5% fee (just like a credit card/payment processor company) and then give them cashback; for example up to 10% in Ubiquity Credits (uCR) or the total of all sum fees (e.g. 2.5% OpenSea, 5% creator fee, 2.5% our processing fee)

> アレクサンダー.eth | U B I Q U I T Y:
this gives us a new revenue source

> アレクサンダー.eth | U B I Q U I T Y:
behind the scenes, we swap the ubiquity dollar for eth/weth used for the purchase and complete the opensea transaction

> T Flex 👀:
i wouldn’t replace the button but add another one

## Smart Contract

The purpose of the "middle-man" contract would be to perform the following:

1. receive the user's uAD
1. set aside uCR valued at 10% of this uAD amount
    1. for security reasons, we should top up this contract manually with uCR
1. check now for the NFT price, and what asset the NFT is priced in (WETH, ETH, DAI, USDC?) and if the uAD amount is valued at 102.5% of the NFT price, not including slippage on the uAD swap.
    1. we should automatically handle choosing and swapping for the correct purchase asset as an extra convenience for the user
1. swap for ETH using the user's uAD (this should be valued at 102.5%+ of the NFT price)
1. purchase NFT with 97.5%, and keep 2.5% of the purchase amount, in the swapped for asset, e.g. ETH
1. send back NFT + uCR "cashback" to user

### Setters

We should be able to set the following variables

1. transaction fee (default at 2.5%)
1. cashback percentage (default at 10%) 
1. account for slippage (default true)

## Remarks

This gives us a reason to do a bond for uAD-ETH and grow its liquidity.

**Comment by 0x4007** [comment]
gm unfortunately opensea item viewer page is down so I couldn't finish the cx
 
<img width="545" alt="image" src="https://user-images.githubusercontent.com/4975670/169495452-81fe0bb4-81c3-4594-b03a-e7578c36b032.png">

<img width="309" alt="image" src="https://user-images.githubusercontent.com/4975670/169495411-26971e15-2543-4d0d-8727-b911743bbd36.png">

**Reply by 0x4007** [reply]
Updated to dollar icon cause it looks the best in this context, and also its a product for the dollar protocol 

<img width="187" alt="image" src="https://user-images.githubusercontent.com/4975670/169501845-fa258591-786d-44f3-be9b-c3f3add76cf0.png">

**Comment by 0x4007** [comment]
cx injects button but can use a couple of refinements. just need the smart contracts now! 

<img width="1366" alt="Screen Shot 2022-05-21 at 17 23 01" src="https://user-images.githubusercontent.com/4975670/169645069-20facc2d-dcee-4336-a0b6-788d7b826b99.png">

**Comment by 0x4007** [comment]
> The reward asset = Ubiquity Credits (uCR) referred herein as “Credits”. 
Credits are convertible to Dollars via our [first party smart contract](https://etherscan.io/address/0x432120Ad63779897A424f7905BA000dF38A74554) when Dollars trade above peg ($1.00+) in our [Curve metapool](https://etherscan.io/address/0x20955CB69Ae1515962177D164dfC9522feef567E).
This allows Credits holders to print the new supply of Dollars, which they sell to make a profit, pushing the price of Dollars back down to peg ($1.00).
Notice: these were recently rebranded from uAR (Ubiquity Auto Redeem) and uDEBT (Ubiquity Debts) tokens. You will likely see references of these in old documentation and code. We generally used to refer to them all as “debts”. Read the original explainer article [here](https://dao.ubq.fi/a-deep-dive-into-ubiquitys-debt-system).

**Comment by 0x4007** [comment]
I think we should make this work with [genie.xyz](https://genie.xyz) UI. Its a marketplace aggregator so we don't need to recreate the UI for each market place.

[example transactions](https://etherscan.io/address/0x0a267cf51ef038fc00e71801f5a524aec06e4f07)

[GenieSwap.sol](https://etherscan.deth.net/address/0x0a267cf51ef038fc00e71801f5a524aec06e4f07)

**Comment by sergfeldman** [comment]
FYI NFT aggregators
https://www.genie.xyz/ 
https://www.gem.xyz/ 
https://flip.xyz/

**Comment by sergfeldman** [comment]
Gem API doc
https://gemxyz.notion.site/What-is-an-API-How-do-I-apply-for-a-Gem-API-ced433e5349b440eb4049c024dfec435

Gem alternative
https://reservoir.tools/

**Comment by 0x4007** [comment]
Sudoswap adding OHM and APE as a currency in their UI?

https://twitter.com/sudoswap/status/1581163275525373953?s=46&t=HI7AqgFr5xp5xcpzwjUPLw
